### PR TITLE
Electric and Misty Terrain Effects

### PR DESCRIPTION
**Pull request by ghoulslash**
## Description
Adds electric terrain preventing sleep and misty terrain preventing non-volatile status conditions

Also adds the following functions to make status checks easier:
```
bool32 CanSleep(u8 battlerId);
bool32 CanBePoisoned(u8 battlerId);
bool32 CanBeBurned(u8 battlerId);
bool32 CanBeParalyzed(u8 battlerId);
bool32 CanBeFrozen(u8 battlerId);
```

misty_terrain

### CODE REVIEW

**Review by ultima-soul — CHANGES_REQUESTED** [review]
Please merge to resolve conflicts.

Also some issues noted:
```
Electric Terrain:
- When Yawn used before Electric Terrain is set and marks a Pokemon as drowsy:
    - If Electric Terrain set the turn after, instead of falling asleep, "X surrounds itself with electrified terrain." is displayed and drowsiness is unset
- When Yawn, Rest, or sleep-inducing move is used while Electric Terrain is set:
    - "X surrounds itself with electrified terrain." is displayed for targeted Pokemon.
        - For Rest, only if HP isn't full it says this, otherwise HP full message appears
        - For Yawn, only if drowsiness isn't set from prior turn, otherwise "But it failed" prints (Yawn used twice against the same Pokemon. One the prior turn and one turn after while drowsiness is still set on the same turn that Electric Terrain is set.)

Misty Terrain:
- Toxic Spikes shouldn't do anything when mon is sent in while Misty Terrain is active
- See Electric Terrain for Sleep-related issues
```

File: include/battle_util.h [hunk 1]
```diff
@@ -148,4 +148,13 @@ bool32 IsGastroAcidBannedAbility(u16 ability);
...
 
+bool32 CanSleep(u8 battlerId);
+bool32 CanBePoisoned(u8 battlerId);
+bool32 CanBeBurned(u8 battlerId);
+bool32 CanBeParalyzed(u8 battlerId);
+bool32 CanBeFrozen(u8 battlerId);
+bool32 CanBeConfused(u8 battlerId);
+bool32 IsBattlerTerrainAffected(u8 battlerId, u32 terrainFlag);
+
+
```
Comment (ultima-soul) [inline include/battle_util.h:159]:
```suggestion
```

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -2617,29 +2606,14 @@ void SetMoveEffect(bool32 primary, u32 certain)
...
             }
-            if (IS_BATTLER_OF_TYPE(gEffectBattler, TYPE_FIRE))
-                break;
-            if (GetBattlerAbility(gEffectBattler) == ABILITY_WATER_VEIL
-                || GetBattlerAbility(gEffectBattler) == ABILITY_COMATOSE
-                || IsAbilityStatusProtected(gEffectBattler))
-                break;
-            if (gBattleMons[gEffectBattler].status1)
+            
+            if (!CanBeBurned(gEffectBattler))
```
Comment (ultima-soul) [inline src/battle_script_commands.c:2610]:
```suggestion
            if (!CanBeBurned(gEffectBattler))
```

File: src/battle_util.c [hunk 3]
```diff
@@ -2727,11 +2727,24 @@ u8 DoBattlerEndTurnEffects(void)
...
                 {
                     CancelMultiTurnMoves(gActiveBattler);
-                    gBattleMons[gActiveBattler].status1 |= (Random() & 3) + 2;
-                    BtlController_EmitSetMonData(0, REQUEST_STATUS_BATTLE, 0, 4, &gBattleMons[gActiveBattler].status1);
-                    MarkBattlerForControllerExec(gActiveBattler);
                     gEffectBattler = gActiveBattler;
-                    BattleScriptExecute(BattleScript_YawnMakesAsleep);
+                    if (IsBattlerTerrainAffected(gActiveBattler, STATUS_FIELD_ELECTRIC_TERRAIN))
+                    {
+                        gBattleCommunication[MULTISTRING_CHOOSER] = 1;
```
Comment (ultima-soul) [inline src/battle_util.c:2733]: Should add Multi String Constants like other Multi String Arrays.

**Review by ultima-soul — CHANGES_REQUESTED** [review]
Some issues remaining:
```
Electric Terrain:
- When Yawn is used while Electric Terrain is set and drowsiness is set from Yawn being used against target in the previous turn:
    - "But it failed" should display.
- When Rest is used while Electric Terrain is set:
    - "X surrounds itself with electrified terrain!" should display for targeted Pokemon when HP isn't full.
        - Shouldn't say "But it failed" or set the mon to sleep when less than Max HP.

Misty Terrain:
- When Yawn is used while Misty Terrain is set and drowsiness is set from Yawn being used against target in the previous turn:
    - "But it failed" should display.
- When Rest is used while Misty Terrain is set:
    - "X surrounds itself with a protective mist!" should display for targeted Pokemon when HP isn't full.
        - Shouldn't say "But it failed" or set the mon to sleep when less than Max HP.
```

**Comment by ghoulslash** [comment]
ready for re-review

**Review by ultima-soul — APPROVED** [review]

**Pull request merged**